// @qualweb/util/dist/domUtils/domUtils.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const getElementReferencedByHREF_1 = __importDefault(require("./getElementReferencedByHREF"));
const isElementHiddenByCSS_1 = __importDefault(require("./isElementHiddenByCSS"));
const isElementHiddenByCSSAux_1 = __importDefault(require("./isElementHiddenByCSSAux"));
const videoElementHasAudio_1 = __importDefault(require("./videoElementHasAudio"));
const isElementHidden_1 = __importDefault(require("./isElementHidden"));
const isFocusableBrowser_1 = __importDefault(require("../accessibilityUtils/isFocusableBrowser"));
const isElementVisible_1 = __importDefault(require("./isElementVisible"));
const elementIdIsReferenced_1 = __importDefault(require("./elementIdIsReferenced"));
const isElementADescendantOf_1 = __importDefault(require("./isElementADescendantOf"));
const isElementADescendantOfExplicitRole_1 = __importDefault(require("./isElementADescendantOfExplicitRole"));
const getVideoMetadata_1 = __importDefault(require("./getVideoMetadata"));
const elementHasContent_1 = __importDefault(require("./elementHasContent"));
const getTrimmedText_1 = __importDefault(require("./getTrimmedText"));
const objectElementIsNonText_1 = __importDefault(require("./objectElementIsNonText"));
const isHumanLanguage_1 = __importDefault(require("./isHumanLanguage"));
const getTextSize_1 = __importDefault(require("./getTextSize"));
const cache_1 = require("../cache");
class DomUtils {
    static isElementHidden(element) {
        return (0, isElementHidden_1.default)(element);
    }
    static isElementHiddenByCSS(element) {
        return (0, isElementHiddenByCSS_1.default)(element);
    }
    static isElementVisible(element) {
        return (0, isElementVisible_1.default)(element);
    }
    static elementIdIsReferenced(element, id, attribute) {
        return (0, elementIdIsReferenced_1.default)(element, id, attribute);
    }
    static isElementADescendantOf(element, names, roles) {
        return (0, isElementADescendantOf_1.default)(element, names, roles);
    }
    static isElementADescendantOfExplicitRole(element, names, roles) {
        return (0, isElementADescendantOfExplicitRole_1.default)(element, names, roles);
    }
    static elementHasContent(element, checkChildren) {
        return (0, elementHasContent_1.default)(element, checkChildren);
    }
    static isElementHiddenByCSSAux(element) {
        return (0, isElementHiddenByCSSAux_1.default)(element);
    }
    static getTrimmedText(element) {
        return (0, getTrimmedText_1.default)(element);
    }
}
DomUtils.getVideoMetadata = getVideoMetadata_1.default;
DomUtils.getElementReferencedByHREF = getElementReferencedByHREF_1.default;
DomUtils.videoElementHasAudio = videoElementHasAudio_1.default;
DomUtils.isFocusableBrowser = isFocusableBrowser_1.default;
DomUtils.objectElementIsNonText = objectElementIsNonText_1.default;
DomUtils.isHumanLanguage = isHumanLanguage_1.default;
DomUtils.getTextSize = getTextSize_1.default;
__decorate([
    (0, cache_1.Cache)('DomUtils.isElementHidden'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Function]),
    __metadata("design:returntype", Boolean)
], DomUtils, "isElementHidden", null);
__decorate([
    (0, cache_1.Cache)('DomUtils.isElementHiddenByCSS'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Function]),
    __metadata("design:returntype", Boolean)
], DomUtils, "isElementHiddenByCSS", null);
__decorate([
    (0, cache_1.Cache)('DomUtils.isElementVisible'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Function]),
    __metadata("design:returntype", Boolean)
], DomUtils, "isElementVisible", null);
__decorate([
    (0, cache_1.FullMethodCache)('DomUtils.elementIdIsReferenced'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Function, String, String]),
    __metadata("design:returntype", Boolean)
], DomUtils, "elementIdIsReferenced", null);
__decorate([
    (0, cache_1.FullMethodCache)('DomUtils.isElementADescendantOf'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Function, Array,
        Array]),
    __metadata("design:returntype", Boolean)
], DomUtils, "isElementADescendantOf", null);
__decorate([
    (0, cache_1.FullMethodCache)('DomUtils.isElementADescendantOfExplicitRole'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Function, Array,
        Array]),
    __metadata("design:returntype", Boolean)
], DomUtils, "isElementADescendantOfExplicitRole", null);
__decorate([
    (0, cache_1.Cache)('DomUtils.elementHasContent'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Function, Boolean]),
    __metadata("design:returntype", Boolean)
], DomUtils, "elementHasContent", null);
__decorate([
    (0, cache_1.Cache)('DomUtils.isElementHiddenByCSSAux'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Function]),
    __metadata("design:returntype", Boolean)
], DomUtils, "isElementHiddenByCSSAux", null);
__decorate([
    (0, cache_1.Cache)('DomUtils.getTrimmedText'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Function]),
    __metadata("design:returntype", String)
], DomUtils, "getTrimmedText", null);
exports.default = DomUtils;
