// @qualweb/core/dist/lib/Report.object.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Report = void 0;
const evaluation_1 = require("./evaluation");
class Report {
    systemData;
    metadata = {
        passed: 0,
        warning: 0,
        failed: 0,
        inapplicable: 0
    };
    modules = {};
    constructor(systemData) {
        this.systemData = systemData;
    }
    addModuleReport(moduleReport) {
        const module = moduleReport.type;
        if (module === 'counter') {
            this.modules[module] = moduleReport;
        }
        else {
            this.modules[module] = moduleReport;
            for (const verdict of Object.values(evaluation_1.Verdict)) {
                this.metadata[verdict] += this.modules[module]?.metadata[verdict] ?? 0;
            }
        }
    }
    getCopy() {
        return {
            type: 'evaluation',
            system: this.systemData,
            metadata: this.metadata,
            modules: this.modules
        };
    }
}
exports.Report = Report;
