// @qualweb/wcag-techniques/dist/lib/Technique.object.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Technique = void 0;
const evaluation_1 = require("@qualweb/core/evaluation");
const techniques_json_1 = __importDefault(require("./techniques.json"));
class Technique extends evaluation_1.Guideline {
    constructor(translator) {
        super();
        this.translator = translator;
        const technique = techniques_json_1.default[new.target.name];
        technique.metadata.passed = 0;
        technique.metadata.warning = 0;
        technique.metadata.failed = 0;
        technique.metadata.inapplicable = 0;
        technique.metadata.outcome = evaluation_1.Verdict.INAPPLICABLE;
        technique.results = new Array();
        this.technique = technique;
        this.translator.translateAssertion(this.technique);
    }
    translate(resultCode, values) {
        return this.translator.translateTest(this.technique.code, resultCode, values);
    }
    getCode() {
        return this.technique.code;
    }
    getMapping() {
        return this.technique.mapping;
    }
    hasPrincipleAndLevels(principles, levels) {
        return this.technique.metadata['success-criteria'].some((sc) => principles.includes(sc.principle) && levels.includes(sc.level));
    }
    getFinalResults() {
        this.generateOutcome();
        return this.technique;
    }
    addTestResult(test) {
        if (!test.description || test.description.trim() === '') {
            test.description = this.translate(test.resultCode);
        }
        this.technique.results.push(test);
        if (test.verdict && test.verdict !== evaluation_1.Verdict.INAPPLICABLE) {
            this.technique.metadata[test.verdict]++;
        }
    }
    generateOutcome() {
        if (this.technique.metadata.failed) {
            this.technique.metadata.outcome = evaluation_1.Verdict.FAILED;
        }
        else if (this.technique.metadata.warning) {
            this.technique.metadata.outcome = evaluation_1.Verdict.WARNING;
        }
        else if (this.technique.metadata.passed) {
            this.technique.metadata.outcome = evaluation_1.Verdict.PASSED;
        }
        else {
            this.technique.metadata.outcome = evaluation_1.Verdict.INAPPLICABLE;
            this.technique.metadata.inapplicable = 1;
        }
        if (this.technique.results.length > 0) {
            this.addDescription();
        }
    }
    addDescription() {
        var _a;
        for (const result of (_a = this.technique.results) !== null && _a !== void 0 ? _a : []) {
            if (result.verdict === this.technique.metadata.outcome) {
                this.technique.metadata.description = result.description;
                break;
            }
        }
    }
}
exports.Technique = Technique;
