// @qualweb/core/dist/lib/evaluation/EvaluationModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EvaluationModuleDefinition = void 0;
const locale_1 = require("@qualweb/locale");
class EvaluationModuleDefinition {
    type;
    moduleOptions;
    report;
    tester;
    translate;
    constructor(type, moduleOptions, translationOptions, report, tester) {
        this.type = type;
        this.moduleOptions = moduleOptions;
        this.report = report;
        this.tester = tester;
        this.translate = locale_1.LocaleFetcher.transform(translationOptions);
        this.configure(moduleOptions);
    }
    configure(options) {
        this.tester.resetConfiguration();
        this.tester.configureByPrinciplesAndLevels(options.principles, options.levels);
        this.tester.configureIncluded(options.include);
        this.tester.configureExcluded(options.exclude);
        return this;
    }
    test(data) {
        this.tester.execute(data);
        return this;
    }
    testSpecial() {
        throw new Error('Method not implemented.');
    }
    getReport() {
        return this.report.getCopy();
    }
}
exports.EvaluationModuleDefinition = EvaluationModuleDefinition;
