// @qualweb/core/dist/lib/evaluation/Test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Test = void 0;
const Verdict_1 = require("./Verdict");
class Test {
    verdict;
    description;
    resultCode;
    elements;
    attributes;
    constructor(verdict, description, resultCode) {
        this.verdict = verdict ?? Verdict_1.Verdict.INAPPLICABLE;
        this.description = description ?? '';
        this.resultCode = resultCode ?? 'I1';
        this.elements = [];
        this.attributes = [];
    }
    addElement(element, withText = true, fullElement = false, aName) {
        const htmlCode = element.getElementHtmlCode(withText, fullElement);
        const pointer = element.getElementSelector();
        let accessibleName;
        if (aName) {
            accessibleName = window.AccessibilityUtils.getAccessibleName(element);
        }
        this.elements.push({ htmlCode, pointer, accessibleName });
        return this;
    }
    addElements(elements, withText = true, fullElement = false, aName) {
        elements.forEach((element) => this.addElement(element, withText, fullElement, aName));
        return this;
    }
}
exports.Test = Test;
